### PR TITLE
tests: replace pthreads with std::thread

### PR DESCRIPTION
**Pull request by marcinslusarz**
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemfile/116)

### CODE REVIEW

**Comment by krzycz** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/pmemfile/116#-:-Ki_YJWcDH44WDBtFINN:bnfp4nl)*

File: tests/posix/mt/mt.cpp [hunk 1]
```diff
@@ -44,13 +44,12 @@ static PMEMfilepool *global_pfp;
 
 class mt : public pmemfile_test {
 public:
-	long ncpus;
-	pthread_t *threads;
+	unsigned ncpus;
+	std::list<std::thread *> threads;
```
Comment (GBuella) [inline tests/posix/mt/mt.cpp]: Why pointers?

File: tests/posix/mt/mt.cpp [hunk 2]
```diff
@@ -187,18 +161,14 @@ TEST_F(mt, pread)
...
-				     (void *)file);
-		ASSERT_EQ(ret, 0) << strerror(errno);
-	}
+	for (unsigned j = 0; j < ncpus; ++j)
+		threads.push_back(new std::thread(pread_worker, file));
 
-	while (i > 0) {
-		ret = pthread_join(threads[--i], NULL);
-		ASSERT_EQ(ret, 0) << strerror(errno);
+	while (!threads.empty()) {
```
Comment (GBuella) [inline tests/posix/mt/mt.cpp]: or:

for (auto& t : threads)
  t.join();

File: tests/posix/mt/mt.cpp (hunk 1 above)
Comment (tomaszkapela) [inline tests/posix/mt/mt.cpp]: I agree with @gbuella, no reason for pointers here. One thing that needs to be mentioned is that std::thread is non-copyable, so remember to either emplace or push r-value references, so they get moved properly.

Comment (marcinslusarz) [inline tests/posix/mt/mt.cpp]: Because I like pointers :P. Fixed.

File: tests/posix/mt/mt.cpp (hunk 2 above)
Comment (marcinslusarz) [inline tests/posix/mt/mt.cpp]: Done.

File: tests/posix/mt/mt.cpp [hunk 3]
```diff
@@ -44,13 +44,12 @@ static PMEMfilepool *global_pfp;
 
 class mt : public pmemfile_test {
 public:
-	long ncpus;
-	pthread_t *threads;
+	unsigned ncpus;
+	std::list<std::thread> threads;
```
Comment (tomaszkapela) [inline tests/posix/mt/mt.cpp]: Just one note from me, remember that std::thread is non-copyable so eplace or push rvalue references so they get moved.

Comment (GBuella) [inline tests/posix/mt/mt.cpp]: Yeah, to make the code more clear, it could be `emplace_back` instead of `push_back`

File: tests/posix/mt/mt.cpp [hunk 4]
```diff
@@ -106,47 +96,29 @@ create_close_unlink_worker(void *arg)
...
-	for (int j = 0; j < ncpus / 2; ++j) {
-		ret = pthread_create(&threads[i++], NULL, open_close_worker,
-				     (void *)"/bbb");
-		ASSERT_EQ(ret, 0) << strerror(errno);
-		ret = pthread_create(&threads[i++], NULL,
-				     create_close_unlink_worker,
-				     (void *)"/bbb");
-		ASSERT_EQ(ret, 0) << strerror(errno);
+	for (unsigned j = 0; j < ncpus / 2; ++j) {
+		threads.push_back(std::thread(open_close_worker, "/bbb"));
```
Comment (GBuella) [inline tests/posix/mt/mt.cpp]: Simpler version:
`threads.emplace_back(open_close_worker, "/bbb");`

File: tests/posix/mt/mt.cpp [hunk 5]
```diff
@@ -106,47 +96,29 @@ create_close_unlink_worker(void *arg)
...
-
-	for (int j = 0; j < ncpus / 2; ++j) {
-		ret = pthread_create(&threads[i++], NULL, open_close_worker,
-				     (void *)"/aaa");
-		ASSERT_EQ(ret, 0) << strerror(errno);
-		ret = pthread_create(&threads[i++], NULL,
-				     create_close_unlink_worker,
-				     (void *)"/aaa");
-		ASSERT_EQ(ret, 0) << strerror(errno);
+	for (unsigned j = 0; j < ncpus / 2; ++j) {
```
Comment (GBuella) [inline tests/posix/mt/mt.cpp]: Are we sure this should be limited by number of cpus?
There can be race issues with multiple threads on a single core as well, as I recall.

File: tests/posix/mt/mt.cpp (hunk 4 above)
Comment (marcinslusarz) [inline tests/posix/mt/mt.cpp]: Done.

File: tests/posix/mt/mt.cpp (hunk 3 above)
Comment (marcinslusarz) [inline tests/posix/mt/mt.cpp]: Done.

File: tests/posix/mt/mt.cpp (hunk 5 above)
Comment (marcinslusarz) [inline tests/posix/mt/mt.cpp]: Take a closer look. It creates twice as many threads as there are cpus.

Comment (GBuella) [inline tests/posix/mt/mt.cpp]: Twice as many?
`j < ncpus / 2`
For example, if `ncpus == 1` this seems to result in:
`for (unsigned j = 0; j < 0; ++j)`

Comment (marcinslusarz) [inline tests/posix/mt/mt.cpp]: Yes. ncpus==1 is the corner case. It was like that before, so i'll make a separate change.
